File: LeetCode-Cpp/N-QueensII.cpp
class Solution {
public:
    int totalNQueens(int n) {
        int count = 0;
        vector<bool> cols(n, false);               // Sütunlar
        vector<bool> diag(2 * n - 1, false);       // Sol üst - sağ alt çaprazlar (row - col + n - 1)
        vector<bool> antiDiag(2 * n - 1, false);   // Sağ üst - sol alt çaprazlar (row + col)

        function<void(int)> backtrack = [&](int row) {
            if (row == n) {
                count++;
                return;
            }

            for (int col = 0; col < n; ++col) {
                int d = row - col + n - 1;
                int ad = row + col;

                if (cols[col] || diag[d] || antiDiag[ad])
                    continue;

                // Veziri yerleştir
                cols[col] = diag[d] = antiDiag[ad] = true;

                backtrack(row + 1);

                // Geri al (backtrack)
                cols[col] = diag[d] = antiDiag[ad] = false;
            }
        };

        backtrack(0);
        return count;

    }
};

File: LeetCode-Python/N-QueensII.py
class Solution:
    def totalNQueens(self, n: int) -> int:
        def backtrack(row: int, diagonals: set, anti_diagonals: set, cols: set) -> int:
            if row == n:
                return 1  # Tüm satırlara başarıyla yerleştirildi

            count = 0
            for col in range(n):
                curr_diag = row - col
                curr_anti_diag = row + col

                # Eğer bu sütun veya çaprazlar zaten doluysa atla
                if col in cols or curr_diag in diagonals or curr_anti_diag in anti_diagonals:
                    continue

                # Veziri yerleştir
                cols.add(col)
                diagonals.add(curr_diag)
                anti_diagonals.add(curr_anti_diag)

                # Bir sonraki satıra geç
                count += backtrack(row + 1, diagonals, anti_diagonals, cols)

                # Geri al (backtrack)
                cols.remove(col)
                diagonals.remove(curr_diag)
                anti_diagonals.remove(curr_anti_diag)

            return count

        return backtrack(0, set(), set(), set())
